lib/Echidna/Exec.hs

```diff
@@ -106,62 +140,65 @@ instance HTraversable VMAction where
 
 
 eCommandCoverage :: (MonadGen n, MonadTest m, MonadState VM m, MonadReader CoverageRef m, MonadIO m)
-                 => (VM -> Bool) -> [SolSignature] -> Command n m VMState
-eCommandCoverage = eCommandUsing $ \(Call c) -> execCallCoverage c
+                 => [SolCall] -> (VM -> Bool) -> [SolSignature] -> [Command n m VMState]
+eCommandCoverage cov p ts = case cov of
+  [] -> [eCommandUsing (genInteractions ts) (\(Call c) -> execCallCoverage c) p]
+  xs -> map (\x -> eCommandUsing (mutateCall x) (\(Call c) -> execCallCoverage c) p) xs
 
 ePropertyUsing :: (MonadCatch m, MonadTest m)
-             => Command Gen m VMState
+             => [Command Gen m VMState]
              -> (m () -> PropertyT IO ())
              -> VM          
              -> Int        
              -> Property
-ePropertyUsing c f v n = mapConfig (\x -> x {propertyTestLimit = 10000}) . property $
+ePropertyUsing cs f v n = mapConfig (\x -> x {propertyTestLimit = 10000}) . property $
   f . executeSequential (VMState v) =<< forAllWith printCallSeq
-  (sequential (linear 1 n) (VMState v) [c]) where
-    printCallSeq = ("Call sequence: " ++) . intercalate "\n               " .
-      map showCall . sequentialActions
-    showCall (Action i _ _ _ _ _) = show i ++ ";"
+  (sequential (linear 1 n) (VMState v) cs)
+  where printCallSeq = ("Call sequence: " ++) . intercalate "\n               " .
+          map showCall . sequentialActions
+        showCall (Action i _ _ _ _ _) = show i ++ ";"
 
 
 ePropertySeq :: (VM -> Bool)   -- Predicate to fuzz for violations of
              -> [SolSignature] -- Type signatures to fuzz
              -> VM             -- Initial state
              -> Int            -- Max actions to execute
              -> Property
-ePropertySeq p ts = ePropertyUsing (eCommand p ts) id             
+ePropertySeq p ts = ePropertyUsing [eCommand (genInteractions ts) p] id             
 
 
-ePropertySeqCoverage :: MVar Coverage
+ePropertySeqCoverage :: [SolCall]
+                     -> MVar [CoverageInfo]
                      -> (VM -> Bool)
                      -> [SolSignature]
                      -> VM
                      -> Int
                      -> Property
-ePropertySeqCoverage globalCovRef p ts v = ePropertyUsing (eCommandCoverage p ts) writeCoverage v where
-  writeCoverage :: MonadIO m => ReaderT CoverageRef (StateT VM m) a -> m a
-  writeCoverage m = do
-    threadCovRef <- liftIO $ newIORef mempty
-    let s = runReaderT m threadCovRef
-    a            <- evalStateT s v
-    threadCov    <- liftIO $ readIORef threadCovRef
-    globalCov    <- liftIO $ takeMVar globalCovRef
-    liftIO $ putMVar globalCovRef (union threadCov globalCov)
-    return a
+ePropertySeqCoverage calls cov p ts v = ePropertyUsing (eCommandCoverage calls p ts) writeCoverage v
+  where
+    writeCoverage :: MonadIO m => ReaderT CoverageRef (StateT VM m) a -> m a
+    writeCoverage m = do
+      threadCovRef <- liftIO $ newIORef (("",[]),empty)
```

Comment:
we defined this on line 56 as emptyCoverageInfo, let's either delete that or use it here